### PR TITLE
The Great Fix of object store acceptance tests ....

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
This PR holds a great variety of fixes which are needed to pass our acceptance test suites.

## Related Issue
refs https://github.com/owncloud/files_primary_s3/pull/48

## Motivation and Context
Acceptance tests should run properly with objectstore as well

## How Has This Been Tested?
- manually
- ci on https://github.com/owncloud/files_primary_s3/pull/48

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature [hunk 1]
```diff
@@ -191,6 +193,7 @@ Feature: trashbin-new-endpoint
 
 	@local_storage
 	@no_default_encryption
+    @skip-on-objectstore
```
[the remark this answers is not in the document]
Comment (phil-davis) [reply tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature]: And use one of the "standards" for naming:
API acceptance tests have made tag names like ``skip_on_objectstore``
webUI acceptance tests have made camelCase tag name like ``skipOnLDAP``
hmmm - it's a shame these diverging "standards" happen.

[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature]: > indentation

github display issue from my pov - I have 4 spaces locally ....

[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature]: underscores are so damn ugly ... I hate them in each and every owncloud app name .....

[the remark this answers is not in the document]
Comment (patrickjahns) [reply tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature]: > hmmm - it's a shame these diverging "standards" happen. 

we can refactor and unify it again - if we have a standard we can also document it and ask people to follow it. Worst case, have a gherkin "code-style" in place

[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply tests/acceptance/features/apiTrashbin/trashbin-new-endpoint.feature]: adopted to existing standard with _

File: lib/private/Files/ObjectStore/NoopScanner.php [hunk 2]
```diff
@@ -58,7 +58,7 @@ public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData =
 	public function scan($path, $recursive = self::SCAN_RECURSIVE, $reuse = -1, $lock = true) {
 		// we only update the checksums - still returning no data
 		$meta = $this->storage->getMetaData($path);
-		if (isset($meta['checksum'])) {
+		if (isset($meta['checksum']) && !empty($meta['checksum'])) {
```
Comment (butonic) [inline lib/private/Files/ObjectStore/NoopScanner.php]: redundant, only  `!empty` is sufficient, see https://stackoverflow.com/a/4559976

**Review by butonic — CHANGES_REQUESTED** [review]

File: lib/private/Files/ObjectStore/NoopScanner.php (hunk 2 above)
Comment (patrickjahns) [reply lib/private/Files/ObjectStore/NoopScanner.php]: Not exactly - different behaviors: https://www.virendrachandak.com/techtalk/php-isset-vs-empty-vs-is_null/

File: lib/private/Share20/DefaultShareProvider.php [hunk 3]
```diff
@@ -669,7 +669,8 @@ private function isAccessibleResult($data) {
 		// exclude shares leading to trashbin on home storages
 		$pathSections = \explode('/', $data['path'], 2);
 		// FIXME: would not detect rare md5'd home storage case properly
-		if ($pathSections[0] !== 'files' && \explode(':', $data['storage_string_id'], 2)[0] === 'home') {
+		$storagePrefix = \explode(':', $data['storage_string_id'], 2)[0];
+		if ($pathSections[0] !== 'files' && \in_array($storagePrefix, ['home', 'object'], true)) {
```
Comment (mrow4a) [inline lib/private/Share20/DefaultShareProvider.php]: Why object?

File: lib/private/User/User.php [hunk 4]
```diff
@@ -236,6 +236,7 @@ public function delete() {
 
 		// Delete the users entry in the storage table
 		Storage::remove('home::' . $this->getUID());
+		Storage::remove('object::user:' . $this->getUID());
```
Comment (mrow4a) [inline lib/private/User/User.php]: so there are now two homes for user?

File: lib/private/Files/Cache/Propagator.php [hunk 5]
```diff
@@ -93,9 +93,9 @@ public function propagateChange($internalPath, $time, $sizeDifference = 0) {
 				->where($builder->expr()->eq('storage', $builder->createNamedParameter($storageId, IQueryBuilder::PARAM_INT)))
 				->andWhere($builder->expr()->in('path_hash', $hashParams))
 				->andWhere($builder->expr()->gt('size', $builder->expr()->literal(-1, IQueryBuilder::PARAM_INT)));
-		}
 
-		$builder->execute();
+			$builder->execute();
```
Comment (mrow4a) [inline lib/private/Files/Cache/Propagator.php]: :>

File: apps/dav/lib/DAV/CopyPlugin.php [hunk 6]
```diff
@@ -83,7 +84,9 @@ function httpCopy(RequestInterface $request, ResponseInterface $response) {
 				$copySuccess = $sourceNode->copy($destinationNode->getFileInfo()->getPath());
 			}
 			if (!$copySuccess) {
+				$destinationNode->acquireLock(ILockingProvider::LOCK_SHARED);
 				$destinationNode->put($sourceNode->get());
```
Comment (mrow4a) [inline apps/dav/lib/DAV/CopyPlugin.php]: `$destinationNode->put` itself is not setting locks ? Need to go deeper, it is a logic of put itself, isnt it?

File: lib/private/Share20/DefaultShareProvider.php (hunk 3 above)
Comment (DeepDiver1975) [reply lib/private/Share20/DefaultShareProvider.php]: There are basically two different types of storage prefixes: 'home' for regular file system and 'object' for objectstore - the later part was missing in some places - fixed in this PR

File: lib/private/User/User.php (hunk 4 above)
Comment (DeepDiver1975) [reply lib/private/User/User.php]: see above

File: lib/private/Files/Cache/Propagator.php (hunk 5 above)
Comment (DeepDiver1975) [reply lib/private/Files/Cache/Propagator.php]: :twisted_rightwards_arrows:

File: apps/dav/lib/DAV/CopyPlugin.php (hunk 6 above)
Comment (DeepDiver1975) [reply apps/dav/lib/DAV/CopyPlugin.php]: search for other places there put is used - the node is locked before

File: lib/private/Share20/DefaultShareProvider.php (hunk 3 above)
Comment (mrow4a) [reply lib/private/Share20/DefaultShareProvider.php]: ok, makes sense!

File: lib/private/User/User.php (hunk 4 above)
Comment (mrow4a) [reply lib/private/User/User.php]: can be something else then X=user in object::X:  ?

File: lib/private/Files/ObjectStore/NoopScanner.php (hunk 2 above)
Comment (butonic) [reply lib/private/Files/ObjectStore/NoopScanner.php]: `isset` checks if the variable, in this case the array key is set, and the value is **not** `null`.
`!empty` does the same thing: `empty` will check if the the array key is set, and the value **is** `null` (or `''`, `'0'`, `0`, `0.0`, `false` or `[]`). the result is negated. it literally checks what `isset` did again.

You could nitpick that `!empty` will return false for `'0'`, `0`, `0.0` ... which may or may not be valid checksums. But I think @DeepDiver1975 just added the empty check to handle `''` correctly. And checksums with `''`, `'0'`, `0`, `0.0`, `null`, `false` or `[]` dont make much sense.

Get rid of the `isset`. it is redundant.

**Review by butonic — APPROVED** [review]

**Comment by patrickjahns** [comment]
Since this PR got merged, I am seeing frequent issues in the logfile 

```
{"reqId":"PvBstjnGLSwoDKQm9Ktg","level":3,"time":"2018-05-18T11:35:12+00:00","remoteAddr":"172.20.0.5","user":"user1","app":"PHP","method":"GET","url":"\/ocs\/v1.php\/cloud\/users\/user1","message":"fclose() expects parameter 1 to be resource, object given at \/var\/www\/owncloud\/lib\/private\/legacy\/util.php#437"}
```

**Comment by mrow4a** [comment]
Can you track down in what case this copyr gets object instead of resource and how to reproduce it ?

**Comment by patrickjahns** [comment]
first guess as it just accesses `/ocs/v1.php/cloud/users/user1` - this might be related to ocs api calls - will push some more objectstorage tests into apps and get steps to reproduce

**Comment by DeepDiver1975** [comment]
copyr is used to copy over the skeleton files on first login ...

**Comment by mrow4a** [comment]
No idea, I have never been in that part of code - it is also legacy - I assumed it fixes some bug, but apparently it makes some tests fail, probably also because of how unit tests are mocking things.

**Comment by phil-davis** [comment]
I fixed other things in the test scenario "Trashbin can be emptied" in master PR https://github.com/owncloud/core/pull/31523
When backporting to stable10, I discovered that I needed the commit:
"Handle CSRF token errors on ajax/delete.php request"
from here.
That was because I  fixed scenario "Trashbin can be emptied" so that it really worked, and had the possibility to fail.
So that commit "Handle CSRF token errors on ajax/delete.php request" has been backported to stable10 as part of https://github.com/owncloud/core/pull/31536

**Comment by phil-davis** [comment]
@DeepDiver1975 what is happening about backporting this PR?
It would be nice to have at least other bits-and-pieces from here backported, if not the whole PR.

**Comment by phil-davis** [comment]
@DeepDiver1975 what is the status of backporting this PR?

This is another one that seems to have a pile of commits gathering..

**Comment by phil-davis** [comment]
I am looking through the listed of clossed PRs that still have backport-request label.
Thankfully most are easy, like this. The backport has been done and just the label needs to be removed.

**Pull request merged**